platform/javascript/detect.py

```diff
@@ -46,38 +49,55 @@ def configure(env):
         # run-time performance.
         env.Append(CCFLAGS=['-Os'])
         env.Append(LINKFLAGS=['-Os'])
-        if env['target'] == 'release_debug':
-            env.Append(CPPDEFINES=['DEBUG_ENABLED'])
-            # Retain function names for backtraces at the cost of file size.
-            env.Append(LINKFLAGS=['--profiling-funcs'])
-    else:
+    elif env['target'] == 'release_debug':
+        env.Append(CCFLAGS=['-Os'])
+        env.Append(LINKFLAGS=['-Os'])
+        env.Append(CPPDEFINES=['DEBUG_ENABLED'])
+        # Retain function names for backtraces at the cost of file size.
+        env.Append(LINKFLAGS=['--profiling-funcs'])
+    else: # 'debug'
         env.Append(CPPDEFINES=['DEBUG_ENABLED'])
         env.Append(CCFLAGS=['-O1', '-g'])
         env.Append(LINKFLAGS=['-O1', '-g'])
         env.Append(LINKFLAGS=['-s', 'ASSERTIONS=1'])
 
-    ## Compiler configuration
+    if env['tools']:
+        if not env['threads_enabled']:
+            raise RuntimeError("Threads must be enabled to build the editor. Please add the 'threads_enabled=yes' option")
+        # Tools need more memory
+        env.Append(LINKFLAGS=['-s', 'TOTAL_MEMORY=33554432'])
+    else:
+        # Disable exceptions and rtti on non-tools (template) builds
+        # These flags help keep the file size down.
+        env.Append(CCFLAGS=['-fno-exceptions', '-fno-rtti'])
+        # Don't use dynamic_cast, necessary with no-rtti.
+        env.Append(CPPDEFINES=['NO_SAFE_CAST'])
 
+    ## Copy env variables.
     env['ENV'] = os.environ
 
-    em_config_file = os.getenv('EM_CONFIG') or os.path.expanduser('~/.emscripten')
-    if not os.path.exists(em_config_file):
-        raise RuntimeError("Emscripten configuration file '%s' does not exist" % em_config_file)
-    with open(em_config_file) as f:
-        em_config = {}
-        try:
-            # Emscripten configuration file is a Python file with simple assignments.
-            exec(f.read(), em_config)
-        except StandardError as e:
-            raise RuntimeError("Emscripten configuration file '%s' is invalid:\n%s" % (em_config_file, e))
-    if 'BINARYEN_ROOT' in em_config and os.path.isdir(os.path.join(em_config.get('BINARYEN_ROOT'), 'emscripten')):
-        # New style, emscripten path as a subfolder of BINARYEN_ROOT
-        env.PrependENVPath('PATH', os.path.join(em_config.get('BINARYEN_ROOT'), 'emscripten'))
-    elif 'EMSCRIPTEN_ROOT' in em_config:
-        # Old style (but can be there as a result from previous activation, so do last)
-        env.PrependENVPath('PATH', em_config.get('EMSCRIPTEN_ROOT'))
-    else:
-        raise RuntimeError("'BINARYEN_ROOT' or 'EMSCRIPTEN_ROOT' missing in Emscripten configuration file '%s'" % em_config_file)
+    # LTO
+    if env['use_lto']:
+        env.Append(CCFLAGS=['-s', 'WASM_OBJECT_FILES=0'])
+        env.Append(LINKFLAGS=['-s', 'WASM_OBJECT_FILES=0'])
+        env.Append(LINKFLAGS=['--llvm-lto',  '1'])
```

Comment:
Might be worth backporting this to the 3.2.x branch too (and anything else that made it possible/safe).